File: python/src/st_quadrant.py
#
# STAR TREK TEXT GAME
# 2017-05-01 K.OHWADA
#

import random


### class	
class Quadrant():
	
	course = None
	
	SIZE_X = 8
	SIZE_Y = 8
	
# easy	
	SETUP_KLINGON = 3	
	
# moderate	
#	SETUP_KLINGON = 5
	
	SETUP_STAR = 5
	# 50 %
	SETUP_PROBABILITY_STARBASE = 0.5
		
	qx =0
	qy =0
	
	# 8x8 Quadrants : klingons, starbases, stars	
	q_arr = []
	
	# 8x8 glaxy record
	g_arr = []

	num_klingon = 0
	num_starbase = 0
	num_star = 0		

	def __init__(self):
		from st_course import Course 
		self.course = Course()
### def end

	def setup_position(self):
		### 1 - 6
		self.qx = int(( self.SIZE_X - 3)*random.random() ) + 1
		self.qy = int(( self.SIZE_Y - 3)*random.random() ) + 1
		

	def setup_q(self):
		self.q_arr = [[ [0,0,0] for j in range(self.SIZE_X)] for i in range(self.SIZE_Y)]
		self.g_arr = [[ "***" for j in range(self.SIZE_Y)] for i in range(self.SIZE_X)]
		
		for i in range(self.SIZE_X):
			for j in range(self.SIZE_Y):
				# setup 3x3 area
				if ( i >= self.qx - 1 ) and ( i <= self.qx + 1 ) and ( j >= self.qy - 1 ) and ( j <= self.qy + 1 ):
					self.q_arr[i][j][0] = int( self.SETUP_KLINGON * random.random() )
					# starbaase 50 %
					self.q_arr[i][j][1] = 1 if (random.random() < self.SETUP_PROBABILITY_STARBASE ) else 0
					self.q_arr[i][j][2] =  int(self.SETUP_STAR * random.random() )												
### def end

### get_current_q
	def get_current_q(self):
		return self.q_arr[self.qx][self.qy]		
### def end	
	
### save_current_q
	def save_current_q(self, q):
		self.q_arr[self.qx][self.qy][0]	 = q[0]
		self.q_arr[self.qx][self.qy][1]	 = q[1]
		self.q_arr[self.qx][self.qy][2]	 = q[2]
				
### def end	

### ount_q
	def count_q(self):
		k = 0
		b = 0
		s = 0
		for i in range(self.SIZE_X):
			for j in range(self.SIZE_Y):				
				k += self.q_arr[i][j][0]
				b += self.q_arr[i][j][1]	
				s += self.q_arr[i][j][2]
### for i j  end

		self.num_klingon = k
		self.num_starbase = b
		self.num_star = s										
### def end	


	
	def print_long_sensor( self, is_computer_available ): 
		map = self.get_map_long(is_computer_available)
		print
		print " LONG RENGE SENSOR"
		self.print_map(map)


### def end	

	def get_map_long(self, is_computer_available):
		map = [[ "" for j in range(self.SIZE_Y)] for i in range(self.SIZE_X)]
		for i in range(self.SIZE_X):
			for j in range(self.SIZE_Y):
				mark = "***"	
				# scan 3x3 area
				if ( i >= self.qx - 1 ) and ( i <= self.qx + 1 ) and ( j >= self.qy - 1 ) and ( j <= self.qy + 1 ):
					info = str(self.q_arr[i][j][0]) + str(self.q_arr[i][j][1]) + str(self.q_arr[i][j][2])
					if (is_computer_available):
						self.g_arr[i][j] = info
					mark = info		
				if (  i == self.qx ) and (  j == self.qy ):
					mark = "E" + mark
### if end	
				map[i][j] = mark
			### for i j end	
		return map
### def end	




	def print_map(self,map):
		self.print_header_row()
		for i in range(self.SIZE_X):
			print i,
			for j in range(self.SIZE_Y):
				mark = map[i][j]
				print mark,
			### for j end
			print
		### for i end			
		msg =  " at Quadrant %d , %d " % (self.qx, self.qy)
		print msg		
### def end
				
	def print_header_row(self):
		print " ",
		for j in range(self.SIZE_Y):
			mark = " " + str(j) + " "
			print mark,
		print	

### def end	
			
	def print_galaxy_record(self):
		print
		print "COMPUTER RECORD OF GALAXY "
		map = self.get_map_galaxy()
		self.print_map(map) 
		
### def end	


	def get_map_galaxy(self):
		map = [[ "" for j in range(self.SIZE_Y)] for i in range(self.SIZE_X)]
		for i in range(self.SIZE_X):
			for j in range(self.SIZE_Y):
				mark = self.g_arr[i][j]
				if (  i == self.qx ) and (  j == self.qy ):
					mark = "E" + mark
				map[i][j] = mark
### for i j	
		return map
### def end
					
	def warp(self, course, distance): 
		print "warp"
		d = self.course.get_course_delta(course)
		x = self.qx + d[0 ]*distance
		y = self.qy + d[1]*distance
		if ( x >= 0 ) and ( x <= 7 ) and ( y >= 0 ) and ( y <= 7 ):
			self.qx = x
			self.qy = y
		else:
			self.setup_position()	
### if end
		self.arrive()
	### def end	
	def arrive(self):
		msg = "arive at Quadrant %d , %d" % ( self.qx, self.qy )
		print msg 
	### def end	
						
	def print_q(self): 
		print "print_q"
		for i in range(8):
			print "i=",
			print i,
			print ":",
			for j in range(8):
				print "j=",
				print j,
				print ":",
				print self.q_arr[i][j],
		### def end
		
		### class end

File: javascript/src/js/st_quadrant.js
//
// STAR TREK Text Game
// 2017-06-01 K.OHWADA
//

var Quadrant = function () {
	 
	 this.course = new Course();
	 
	 // 8x8 Quadrant
	 this.q_arr= [];
	 
	 // galsxy record
	 this.g_arr= [];
	 
this.qx = 0;
this.qy = 0;

this.num_klingon = 0;
this.num_starbase = 0;
this.num_star = 0;
			
}

Quadrant.prototype.setup_position = function () {
	// 1 - 6
	this.qx = Math.floor( 5 *　Math.random() ) + 1 ;
	this.qy = Math.floor( 5 *　Math.random() ) + 1;
	
	// alert( "setup_position" + this.qx + ","+ this.qy);
	
} // setup_position


Quadrant.prototype.setup_q = function () {
	
var i = 0;
var j = 0;


for(i=0;i < 8;i++){
			this.q_arr[i] = [];
			this.g_arr[i] = [];
			
	for(j=0;j < 8;j++){
				this.q_arr[i][j] = [];
		this.q_arr[i][j][0] = 0;
		this.q_arr[i][j][1] = 0;
		this.q_arr[i][j][2] = 0;
		this.g_arr[i][j] = "***";

// setup 3x3 area		
 		if (( i >= this.qx - 1) && ( i <= this.qx + 1) && ( j >= this.qy - 1) && ( j <= this.qy + 1)) {
 			
 			// enterprise or 50 %
			if (((i == this.qx) && (j == this.qy)) || (Math.random() < 0.5) ) {
				
		// klingon 1 - 3		
		this.q_arr[i][j][0] = Math.floor( 2 *　Math.random() ) + 1 ;
		
		// starbase 0 or 1, 50 %
		this.q_arr[i][j][1] = ( Math.random() < 0.5 ) ? 1: 0;
		
		// star 1 - 5
		this.q_arr[i][j][2] = Math.floor( 4 *　Math.random() ) + 1 ;
		
		} // if enterprise or 50 %
} // if3x3 area

}} // for i j

} // setup_q

	
Quadrant.prototype.scan = function (is_computer_available) {
	
var ret = "LONG RENGE SENSOR <br />\n"	
var map = this.get_long_map(is_computer_available);
ret += this.build_map( map );

	return ret;
	
} // scan


Quadrant.prototype.galaxy_record = function () {
	
var ret = "CUMULATIVE GALACTIC RECORD <br />\n"	;
var map = this.get_galaxy_map();
ret += this.build_map( map );

	return ret;
	
} // galaxy_record


Quadrant.prototype.get_long_map = function ( is_computer_available ) {
	
var i = 0;
var j = 0;

var map = [] ;
var mark = "";
var info = "";
	

for(i=0;i < 8;i++){
	map[i] = [] ;
	for(j=0;j < 8;j++){
		mark = "***" ;
		
		// scam 3x3 area
 		if (( i >= this.qx - 1) && ( i <= this.qx + 1) && ( j >= this.qy - 1) && ( j <= this.qy + 1)) {

		info = String(this.q_arr[i][j][0] ) + String(this.q_arr[i][j][1] ) + String(this.q_arr[i][j][2] ) ;
		mark = info;
		
		//copy to galaxy record
		if ( is_computer_available ) {
			this.g_arr[i][j] = info;
		} // if is_computer_available
		
} // if scam 3x3 area

// enterprise
	  	if ( (i == this.qx) && (j== this.qy) ) {
	  		mark = "E" + mark;
	  	} // if enterprise
	  	
	  	
	  	map[i][j] = mark;
	  	
}} // for   i j

	return map;
	
} // get_long_map


Quadrant.prototype.get_galaxy_map = function () {
	
var i = 0;
var j = 0;

var map = [] ;
var mark = "";

	
for(i=0;i < 8;i++){
	map[i] = [] ;
	
	for(j=0;j < 8;j++){
		mark = this.g_arr[i][j];


// enterprise
	  	if ( (i == this.qx) && (j== this.qy) ) {
	  		mark = "E" + mark;
	  	} // if enterprise
	  	
	  	
	  	map[i][j] = mark;
	  	
}} // for   i j

	return map;
	
} // get_galaxy_map




Quadrant.prototype.build_map = function (map) {
		
var i = 0;
var j = 0;

var ret = "";

// render
ret += "&nbsp;&nbsp;";
for(j=0;j < 8;j++){
		ret += "&nbsp;" + String(j) + "&nbsp;";
}
ret += "<br/>\n";

for(i=0;i < 8;i++){
	ret += String(i) + "&nbsp;";
	for(j=0;j < 8;j++){


	  	ret += map[i][j] + "&nbsp;";
} // for  j
	ret += "<br/>\n";
		  	
} // for i
ret += "at Quadrant " + String(this.qx) + " , " + String(this.qy) + "<br/>\n";
	return ret;
	
} // bild_map


Quadrant.prototype.count_q = function () {
	
var i = 0;
var j = 0;

var k = 0;
var b = 0;
var s = 0;


for(i=0;i < 8;i++){
	for(j=0;j < 8;j++){
		k += this.q_arr[i][j][0];
		b += this.q_arr[i][j][1];
		s += this.q_arr[i][j][2];
	}} // for i j
	
	this.num_klingon = k;
		this.num_starbase = b;
			this.num_star = s;
			
	} // count			
	
	
	Quadrant.prototype.get_current_q = function () {
	//		alert( this.qx + ","+ this.qy);
		var param = this.q_arr[this.qx][this.qy];
		return param;
	} // get_current_q	
	
	
		Quadrant.prototype.save_current_q = function (param) {
		this.q_arr[this.qx][this.qy][0] = param[0];
		this.q_arr[this.qx][this.qy][1] = param[1];
		this.q_arr[this.qx][this.qy][2] = param[2];
						
	} // save_current_q	
	
	Quadrant.prototype.warp = function (course) {
		if (( course <1 ) || (  course > 8 ) ) {
			return;
		} // if
			
	//	alert("warp ...");
	
	var d = this.course.get_delta(course);
		
				// distance is 1
		var x = this.qx + d[0];
		var y = this.qy + d[1];	

 		if (( x >= 0) && ( x < 8) && ( y >= 0 ) && ( y < 8 )) {			
			this.qx = x;
			this.qy = y;
			return true;
				
		} else { 
		
			this.setup_position();
			return false;
			
		} // if
			
		} // warp	